c/tests/check_grounding_space.c

```diff
@@ -97,12 +103,291 @@ START_TEST (test_replace)
+        }
+    }
+
+    //We never found the atom
+    return false;
+}
+
+bool remove_atom(void* space_ptr, const atom_t* atom) {
+    custom_space_buf* space = space_ptr;
+
+    //QUESTION FOR VITALY: Not sure if it's right to notify on all removes, or just on
+    // sucessful removes.  Assuming all for now
+    space_event_t* event = space_event_new_remove(atom_clone(atom));
+    notify_all_observers(space->observers, event);
+    space_event_free(event);
+
+    return remove_atom_internal(space, atom);
+}
+
+bool replace_atom(void* space_ptr, const atom_t* from, atom_t* to) {
+    custom_space_buf* space = space_ptr;
+
+    //QUESTION FOR VITALY: Not sure if it's right to notify on all replaces, or just on
+    // sucessful ones.  Assuming all for now
+    space_event_t* event = space_event_new_replace(atom_clone(from), atom_clone(to));
+    notify_all_observers(space->observers, event);
+    space_event_free(event);
+
+    if (remove_atom_internal(space, from)) {
+        add_atom_internal(space, to);
+        return true;
+    } else {
+        atom_free(to);
+        return false;
+    }
+}
+
+void free_payload(void* space_ptr) {
+    custom_space_buf* space = space_ptr;
+
+    atom_list_item* cur_atom_item = space->atoms;
+    while (cur_atom_item != NULL) {
+        atom_free(cur_atom_item->atom);
+        atom_list_item* next_item = cur_atom_item->next;
+        free(cur_atom_item);
+        cur_atom_item = next_item;
+    }
+
+    observer_list_item* cur_observer_item = space->observers;
+    while (cur_observer_item != NULL) {
+        space_observer_free(cur_observer_item->observer);
+        observer_list_item* next_item = cur_observer_item->next;
+        free(cur_observer_item);
+        cur_observer_item = next_item;
+    }
+
+    free(space);
+}
+
+START_TEST (test_custom_c_space)
+{
+    space_api_t const C_SPACE_API= { 
+        .register_observer = &register_observer,
+        .query = &query,
+        .subst = NULL,
+        .add = &add_atom,
+        .remove = &remove_atom,
+        .replace = &replace_atom,
+        .free_payload = &free_payload
+    };
+    custom_space_buf* c_space = malloc(sizeof(custom_space_buf));
+    c_space->observers = NULL;
+    c_space->atoms = NULL;
+    c_space->atom_count = 0;
+    space_t* space = space_new(&C_SPACE_API, c_space);
+
+    atom_t* a = atom_sym("A");
+    atom_t* b = atom_sym("B");
+    atom_t* c = atom_sym("C");
+    space_add(space, atom_sym("A"));
+    space_add(space, atom_sym("B"));
+    space_add(space, atom_sym("C"));
+    ck_assert(space_remove(space, a));
+    ck_assert(!space_remove(space, a));
+    ck_assert(space_remove(space, c));
+    ck_assert(space_remove(space, b));
+    space_add(space, atom_sym("A"));
+    ck_assert(space_replace(space, a, atom_sym("B")));
+    ck_assert(!space_replace(space, a, atom_sym("Junk")));
+    ck_assert(space_remove(space, b));
+    atom_free(a);
+    atom_free(b);
+    atom_free(c);
```

Comment:
`a` is created in line 351, then it is moved into `space_add` in line 354, then it is freed. It is not clear to me, why this doesn't lead to the double free? :-)

Reply:
Line 354 says:
```c
    space_add(space, atom_sym("A"));
```
it does not say:
```c
    space_add(space, a);
```
From MeTTa's viewpoint, `a` and atom_sym("A") are the same.  From from C's point of view, they are separate objects, that happen to contain exactly the same data.

Reply:
Thanks, such a big PR, that I have confused myself at the end.